From 336335a78e24b4b0ad43a61caf2df4f2cc15601b Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Thu, 3 Feb 2011 18:29:24 -0700
Subject: [PATCH 4/7] Add a Search list screen and editor for Searches

Add a Search list screen which displays the list of Searches
owned by a specific owner.  Also add a Search editor widget to
make any of the searches in this list editable.

Change-Id: I3403bb1ed2220df40ca9bbefba9deb64d65e0671
---
 .../java/com/google/gerrit/client/Dispatcher.java  |   48 +++
 .../java/com/google/gerrit/client/GerritCss.java   |    1 +
 .../main/java/com/google/gerrit/client/gerrit.css  |    5 +
 .../gerrit/client/searches/SearchConstants.java    |   67 ++++
 .../client/searches/SearchConstants.properties     |   34 ++
 .../gerrit/client/searches/SearchEditor.java       |  339 ++++++++++++++++++++
 .../gerrit/client/searches/SearchMessages.java     |   36 ++
 .../client/searches/SearchMessages.properties      |    3 +
 .../gerrit/client/searches/SearchScreen.java       |  235 ++++++++++++++
 .../google/gerrit/client/searches/SearchTable.java |  290 +++++++++++++++++
 .../com/google/gerrit/client/searches/Util.java    |   64 ++++
 .../com/google/gerrit/client/ui/HintTextBox.java   |    6 +-
 .../google/gerrit/client/ui/NavigationTable.java   |   14 +
 .../gerrit/client/ui/SwitchingSuggestOracle.java   |   76 +++++
 .../gerrit/client/ui/TextBoxBaseListener.java      |  121 +++++++
 15 files changed, 1338 insertions(+), 1 deletions(-)
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.java
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.properties
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchEditor.java
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchMessages.java
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchMessages.properties
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchScreen.java
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchTable.java
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/Util.java
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/SwitchingSuggestOracle.java
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/TextBoxBaseListener.java

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java
index 383a226..7e47cc7 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java
@@ -57,15 +57,18 @@ import com.google.gerrit.client.changes.PatchTable;
 import com.google.gerrit.client.changes.PublishCommentScreen;
 import com.google.gerrit.client.changes.QueryScreen;
 import com.google.gerrit.client.patches.PatchScreen;
+import com.google.gerrit.client.searches.SearchScreen;
 import com.google.gerrit.client.ui.Screen;
 import com.google.gerrit.common.auth.SignInMode;
 import com.google.gerrit.common.data.PatchSetDetail;
 import com.google.gerrit.reviewdb.Account;
 import com.google.gerrit.reviewdb.AccountGroup;
 import com.google.gerrit.reviewdb.Change;
+import com.google.gerrit.reviewdb.Owner;
 import com.google.gerrit.reviewdb.Patch;
 import com.google.gerrit.reviewdb.PatchSet;
 import com.google.gerrit.reviewdb.Project;
+import com.google.gerrit.reviewdb.Search;
 import com.google.gwt.core.client.GWT;
 import com.google.gwt.core.client.RunAsyncCallback;
 import com.google.gwtorm.client.KeyUtil;
@@ -91,6 +94,26 @@ public class Dispatcher {
     return "admin,project," + n.toString() + "," + tab;
   }
 
+  public static String toSearches(final Owner.Type type) {
+    return "searches," + type.getId();
+  }
+
+  public static String toSearches(final Owner.Type type, final String id) {
+    if (id == null || "".equals(id)) {
+      return toSearches(type);
+    }
+    return toSearches(type) + "," + id;
+  }
+
+  public static String toSearches(final Owner.Type type, final String owner,
+    String search) {
+    if (owner == null || "".equals(owner) ||
+        search == null || "".equals(search)) {
+      return toSearches(type, owner + search);
+    }
+    return toSearches(type, owner + "," + search);
+  }
+
   static final String RELOAD_UI = "reload-ui,";
   private static boolean wasStartedByReloadUI;
 
@@ -128,6 +151,9 @@ public class Dispatcher {
     } else if (token.startsWith("project,")) {
       Gerrit.display(token, project(token));
 
+    } else if (token.startsWith("searches,")) {
+      Gerrit.display(token, searches(token));
+
     } else if (SETTINGS.equals(token) //
         || REGISTER.equals(token) //
         || token.startsWith("settings,") //
@@ -300,6 +326,28 @@ public class Dispatcher {
     });
   }
 
+  private static Screen searches(final String token) {
+    String[] toks = token.split(",");
+    Owner.Type type = Owner.Type.USER;
+    String owner = "";
+    String name = "";
+
+    if (toks.length > 1) {
+      type = Owner.Type.forId(toks[1]);
+    }
+    if (toks.length > 2) {
+      if (type == Owner.Type.SITE) {
+        name = toks[2];
+      } else {
+        owner = toks[2];
+      }
+    }
+    if (toks.length > 3) {
+      name = toks[3];
+    }
+    return new SearchScreen(new Search.Key(type, owner, name));
+  }
+
   private static void settings(String token) {
     GWT.runAsync(new AsyncSplit(token) {
       public void onSuccess() {
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritCss.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritCss.java
index ca9f248..d4f286c 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritCss.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritCss.java
@@ -162,6 +162,7 @@ public interface GerritCss extends CssResource {
   String rpcStatusPanel();
   String screen();
   String screenHeader();
+  String searchEditor();
   String searchPanel();
   String sectionHeader();
   String sideBySideScreenLinkTable();
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/gerrit.css b/gerrit-gwtui/src/main/java/com/google/gerrit/client/gerrit.css
index c8f16d0..183710b 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/gerrit.css
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/gerrit.css
@@ -1234,3 +1234,8 @@ a:hover.downloadLink {
   margin: 4px;
   opacity: 0.90;
 }
+
+.searchEditor {
+  background: trimColor;
+  width: 100%;
+}
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.java
new file mode 100644
index 0000000..160facb
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.java
@@ -0,0 +1,67 @@
+// Copyright (C) 2011 The Android Open Source Project
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.client.searches;
+
+import com.google.gwt.i18n.client.Constants;
+
+public interface SearchConstants extends Constants {
+  String defaultSearchName();
+
+  String buttonDeleteSearches();
+  String buttonCancelSearch();
+  String buttonInsertNewSearch();
+  String buttonInsertSearch();
+  String buttonSaveSearch();
+
+  String columnSearchName();
+  String columnSearchDescription();
+  String columnSearchQuery();
+
+  String columnSearchEditName();
+  String columnSearchEditDescription();
+  String columnSearchEditQuery();
+  String columnSearchEditType();
+  String columnSearchEditGroup();
+  String columnSearchEditProject();
+
+  String ownerTypeUser();
+  String ownerTypeGroup();
+  String ownerTypeProject();
+  String ownerTypeSite();
+
+  String searchListPrev();
+  String searchListNext();
+  String searchListOpen();
+
+  String titleSearchUser();
+  String titleSearchGroup();
+  String titleSearchProject();
+  String titleSearchSite();
+
+  String searchTableNone();
+}
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.properties b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.properties
new file mode 100644
index 0000000..430d9bd
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.properties
@@ -0,0 +1,34 @@
+defaultSearchName = (Untitled)
+
+buttonDeleteSearches = Delete
+buttonCancelSearch = Cancel Edit
+buttonInsertNewSearch = Add New Search
+buttonInsertSearch = Add Search
+buttonSaveSearch = Save Search
+
+columnSearchName = Name
+columnSearchDescription = Description
+columnSearchQuery = Query String
+
+columnSearchEditName = Name:
+columnSearchEditDescription = Description:
+columnSearchEditQuery = Query String:
+columnSearchEditType = Search Type:
+columnSearchEditGroup = Group Name:
+columnSearchEditProject = Project Name:
+
+ownerTypeUser = User
+ownerTypeGroup = Group
+ownerTypeProject = Project
+ownerTypeSite = Site
+
+searchListPrev = Previous search
+searchListNext = Next search
+searchListOpen = Open search
+
+titleSearchUser = User Searches
+titleSearchGroup = Group Searches
+titleSearchProject = Project Searches
+titleSearchSite = Site Searches
+
+searchTableNone = (None)
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchEditor.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchEditor.java
new file mode 100644
index 0000000..6cfb83d
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchEditor.java
@@ -0,0 +1,339 @@
+// Copyright (C) 2011 The Android Open Source Project
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.client.searches;
+
+import com.google.gerrit.client.Gerrit;
+import com.google.gerrit.client.rpc.GerritCallback;
+import com.google.gerrit.client.ui.AccountGroupSuggestOracle;
+import com.google.gerrit.client.ui.HintTextBox;
+import com.google.gerrit.client.ui.OnEditEnabler;
+import com.google.gerrit.client.ui.ProjectNameSuggestOracle;
+import com.google.gerrit.client.ui.RPCSuggestOracle;
+import com.google.gerrit.client.ui.SwitchingSuggestOracle;
+import com.google.gerrit.client.ui.TextBoxBaseListener;
+import com.google.gerrit.common.data.SearchList;
+import com.google.gerrit.reviewdb.Owner;
+import com.google.gerrit.reviewdb.Search;
+
+import com.google.gwt.event.dom.client.ChangeEvent;
+import com.google.gwt.event.dom.client.ChangeHandler;
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.user.client.ui.Button;
+import com.google.gwt.user.client.ui.Composite;
+import com.google.gwt.user.client.ui.FlowPanel;
+import com.google.gwt.user.client.ui.Grid;
+import com.google.gwt.user.client.ui.Label;
+import com.google.gwt.user.client.ui.ListBox;
+import com.google.gwt.user.client.ui.SuggestBox;
+import com.google.gwt.user.client.ui.SuggestOracle;
+import com.google.gwtexpui.globalkey.client.NpTextArea;
+import com.google.gwtexpui.globalkey.client.NpTextBox;
+
+public class SearchEditor extends Composite {
+  private HintTextBox name;
+  private NpTextArea description;
+  private HintTextBox query;
+  private NpTextBox owner;
+  private SuggestBox ownerSug;
+  private SwitchingSuggestOracle oracle;
+  private ListBox type;
+
+  private Button save;
+  private Button insert;
+  private Button cancel;
+
+  private OnEditEnabler saveEnabler;
+  private OnEditEnabler insertEnabler;
+
+  private final int nrows = 5;
+  private Grid grid = new Grid(nrows, 2);
+
+  private SearchList list;
+  private Search search;
+  private int r_owner;
+
+  public SearchEditor(Search search) {
+    createWidgets();
+
+    setupSaveEnablers();
+
+    name.setVisibleLength(45);
+    description.setWidth("99%");
+    query.setWidth("99%");
+    owner.setVisibleLength(45);
+
+    int r = 0, c = 0;
+    grid.setText(r++, c, Util.C.columnSearchEditName());
+    grid.setText(r++, c, Util.C.columnSearchEditDescription());
+    grid.setText(r++, c, Util.C.columnSearchEditQuery());
+    grid.setText(r++, c, Util.C.columnSearchEditType());
+    grid.setText(r++, c, "");
+
+    r = 0; c++;
+    grid.setWidget(r++, c, name);
+    grid.setWidget(r++, c, description);
+    grid.setWidget(r++, c, query);
+    grid.setWidget(r++, c, type);
+    r_owner = r;
+    grid.setWidget(r++, c, new Label(""));
+    grid.setStyleName(Gerrit.RESOURCES.css().searchEditor());
+
+
+    final FlowPanel buttons = new FlowPanel();
+    buttons.add(save);
+    buttons.add(insert);
+    buttons.add(cancel);
+
+
+    final FlowPanel fp = new FlowPanel();
+    fp.add(grid);
+    fp.add(buttons);
+    setWidget(fp);
+
+    setSearch(search);
+  }
+
+  public void createWidgets() {
+    save = new Button(Util.C.buttonSaveSearch());
+    save.addClickHandler(new ClickHandler() {
+      @Override
+      public void onClick(final ClickEvent event) {
+        onSave(false);
+      }
+    });
+
+    insert = new Button(Util.C.buttonInsertSearch());
+    insert.addClickHandler(new ClickHandler() {
+      @Override
+      public void onClick(final ClickEvent event) {
+        onSave(true);
+      }
+    });
+
+    cancel = new Button(Util.C.buttonCancelSearch());
+    cancel.addClickHandler(new ClickHandler() {
+      @Override
+      public void onClick(final ClickEvent event) {
+        onCancel();
+      }
+    });
+
+    name = new HintTextBox();
+    name.setHintText(Util.C.defaultSearchName());
+
+    TextBoxBaseListener validator = new TextBoxBaseListener(name) {
+      @Override
+      public void onTextChange(String old, String val) {
+        if (! name.isHintOn() && ! Search.Key.isValidName(val)
+            && ! "".equals(val)) { // Need to let users delete all chars
+          name.setText(old);
+        }
+      }
+    };
+
+    description = new NpTextArea();
+    query = new HintTextBox();
+    type = new ListBox();
+    for (Owner.Type t : Owner.Type.values()) {
+      type.addItem(t.getId(), t.getId());
+    }
+    type.addChangeHandler(new ChangeHandler() {
+      @Override
+      public void onChange(ChangeEvent event) {
+        onTypeChange();
+      }
+    });
+    owner = new NpTextBox();
+
+    SuggestOracle[] oracles = new SuggestOracle[2];
+    oracles[0] = new AccountGroupSuggestOracle();
+    oracles[1] = new ProjectNameSuggestOracle();
+    oracle = new SwitchingSuggestOracle(oracles);
+    ownerSug = new SuggestBox(new RPCSuggestOracle(oracle), owner);
+
+    saveEnabler = new OnEditEnabler(save);
+    insertEnabler = new OnEditEnabler(insert);
+ }
+
+  public void onTypeChange() {
+    displayOwner();
+    displayButtons();
+  }
+
+  public void setSearch(Search s) {
+    search = s;
+    if (s == null) {
+      return;
+    }
+
+    name.setText(s.getName());
+    description.setText(s.getDescription());
+    query.setText(s.getQuery());
+    displayType(s.getType() != null ? s.getType() : Owner.Type.USER);
+
+    owner.setText(getOwnerForDisplay());
+    displayOwner();
+
+    save.setEnabled(false);
+    displayButtons();
+  };
+
+  protected Owner.Type getType() {
+    return Owner.Type.forId(type.getValue(type.getSelectedIndex()));
+  }
+
+  public void setSearchList(SearchList list) {
+    this.list = list;
+    displayType(list.getType());
+    setSearch(null);
+  }
+
+  protected String getOwnerForDisplay() {
+    if (list != null) {
+      return list.getOwnerInfo().getOwnerForDisplay();
+    }
+    return "";
+  }
+
+  public void setupSaveEnablers() {
+    saveEnabler.listenTo(name);
+    saveEnabler.listenTo(description);
+    saveEnabler.listenTo(query);
+    saveEnabler.listenTo(owner);
+
+    insertEnabler.listenTo(name);
+    insertEnabler.listenTo(owner);
+  }
+
+  protected void displayOwner() {
+    switch(getType()) {
+      case USER:
+      case SITE:
+          grid.setText(r_owner, 0, "");
+          grid.setText(r_owner, 1, "");
+        break;
+      case GROUP:
+          grid.setText(r_owner, 0, Util.C.columnSearchEditGroup());
+          grid.setWidget(r_owner, 1, ownerSug);
+          oracle.setIndex(0);
+        break;
+      case PROJECT:
+          grid.setText(r_owner, 0, Util.C.columnSearchEditProject());
+          grid.setWidget(r_owner, 1, ownerSug);
+          oracle.setIndex(1);
+        break;
+    }
+
+    if (search.getType() != getType()) {
+      if (search != null && getOwnerForDisplay().equals(owner.getText())) {
+        owner.setText("");
+      }
+    } else if ("".equals(owner.getText())) {
+      owner.setText(getOwnerForDisplay());
+    }
+  }
+
+  protected void displayButtons() {
+    cancel.setEnabled(true);
+    if (search.getKey() == null || ! search.getKey().isValid()) {
+      save.setVisible(false);
+      insert.setEnabled(true);
+    } else if (list != null && ! list.isEditable()) {
+      save.setVisible(false);
+      insert.setEnabled(false);
+      // How could the insertEnabler be disabled here until the type changes?
+    } else {
+      save.setVisible(search.getType() == getType());
+      if (search.getType() == getType()) {
+        insert.setEnabled("".equals(search.getName()));
+      } else {
+        insert.setEnabled(true);
+      }
+    }
+  }
+
+  protected void displayType(Owner.Type t) {
+    int i = 0;
+    for (Owner.Type st : Owner.Type.values()) {
+      if (st.equals(t)) {
+        type.setSelectedIndex(i);
+      }
+      i++;
+    }
+  }
+
+  public void onSave(boolean insert) {
+    save.setEnabled(false);
+    this.insert.setEnabled(false);
+    cancel.setEnabled(false);
+
+    String id = owner.getText();
+    if (getType() == Owner.Type.USER || getType() == Owner.Type.SITE) {
+      id = "";
+    } else if (getOwnerForDisplay().equals(id)) {
+      // When the owner doesn't change, keep the same id on "updated"
+      // search so that others may detect changes based on the id.
+      id = search.getOwner();
+    }
+
+    final Search.Key key = new Search.Key(getType(), id, name.getText());
+    final Search updated = new Search(key);
+
+    updated.setDescription(description.getText());
+    updated.setQuery(query.getText());
+
+    final GerritCallback cb = new GerritCallback<String>() {
+      @Override
+      public void onFailure(Throwable caught) {
+        save.setEnabled(true);
+        cancel.setEnabled(true);
+        super.onFailure(caught);
+      }
+
+      @Override
+      public final void onSuccess(String linkName) {
+        onSaveSuccess(updated, linkName);
+      }
+    };
+
+    if (insert || search.getKey() == null || ! search.getKey().isValid()) {
+      Util.SEARCH_SVC.createSearch(updated, cb);
+    } else {
+      Util.SEARCH_SVC.changeSearch(search.getKey(), updated, cb);
+    }
+  }
+
+  public void onSaveSuccess(Search updated, String linkName) {
+    setSearch(null);
+  }
+
+  public void onCancel() {
+    setSearch(null);
+  }
+}
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchMessages.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchMessages.java
new file mode 100644
index 0000000..7670d6d
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchMessages.java
@@ -0,0 +1,36 @@
+// Copyright (C) 2011 The Android Open Source Project
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.client.searches;
+
+import com.google.gwt.i18n.client.Messages;
+
+public interface SearchMessages extends Messages {
+  String titleSearchUser(String name);
+  String titleSearchGroup(String name);
+  String titleSearchProject(String name);
+}
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchMessages.properties b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchMessages.properties
new file mode 100644
index 0000000..55a99bd
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchMessages.properties
@@ -0,0 +1,3 @@
+titleSearchUser = User Searches for {0}
+titleSearchGroup = Group Searches for {0}
+titleSearchProject = Project Searches for {0}
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchScreen.java
new file mode 100644
index 0000000..50bbd90
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchScreen.java
@@ -0,0 +1,235 @@
+// Copyright (C) 2011 The Android Open Source Project
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.client.searches;
+
+import com.google.gerrit.client.Dispatcher;
+import com.google.gerrit.client.Gerrit;
+import com.google.gerrit.client.ui.Screen;
+import com.google.gerrit.client.rpc.GerritCallback;
+import com.google.gerrit.client.rpc.ScreenLoadCallback;
+import com.google.gerrit.common.data.OwnerInfo;
+import com.google.gerrit.common.data.SearchList;
+import com.google.gerrit.reviewdb.Search;
+
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.user.client.History;
+import com.google.gwt.user.client.ui.Button;
+
+import java.util.List;
+
+public class SearchScreen extends Screen {
+  protected SearchList list;
+  protected SearchTable table;
+  protected Search.Key key;
+
+  protected Button delete;
+  protected Button insert;
+
+  protected ScreenLoadCallback<List<Search>> cbSearches =
+      new ScreenLoadCallback<List<Search>>(this) {
+        @Override
+        protected void preDisplay(final List<Search> searches) {
+          displaySearches(searches);
+        }
+
+        @Override
+        protected void postDisplay() {
+        }
+      };
+
+  protected ScreenLoadCallback<SearchList> cbList =
+      new ScreenLoadCallback<SearchList>(this) {
+        @Override
+        protected void preDisplay(final SearchList list) {
+          display(list);
+        }
+
+        @Override
+        protected void postDisplay() {
+        }
+      };
+
+  protected GerritCallback cbDeletes = new GerritCallback() {
+      public final void onSuccess(Object obj) {
+        loadSearches();
+      }
+    };
+
+  public SearchScreen(final Search.Key key) {
+    super();
+    this.key = key;
+  }
+
+  @Override
+  protected void onInitUI() {
+    super.onInitUI();
+
+    delete = new Button(Util.C.buttonDeleteSearches());
+    delete.addClickHandler(new ClickHandler() {
+        public void onClick(ClickEvent event) {
+          onDelete();
+        }
+      });
+    delete.setVisible(false);
+    delete.setEnabled(false);
+
+    insert = new Button(Util.C.buttonInsertNewSearch());
+    insert.addClickHandler(new ClickHandler() {
+      public void onClick(ClickEvent event) {
+        insert.setVisible(false);
+        table.showEditor(-1);
+      }
+    });
+    insert.setVisible(false);
+
+    table = new SearchTable() {
+      @Override
+      public void onSearchChecked(boolean checked) {
+        delete.setEnabled(checked);
+      }
+
+      @Override
+      protected SearchEditor createSearchEditor() {
+        return new Editor(null) {
+          @Override
+          public void setSearch(Search s) {
+            if (Gerrit.isSignedIn()) {
+            super.setSearch(s);
+              if (list == null) {
+                insert.setVisible(false);
+              } else {
+                insert.setVisible(s == null && list.isEditable());
+              }
+            }
+          }
+
+          @Override
+          public void onSaveSuccess(Search updated, String linkName) {
+            if (updated.getType() == key.getType()) {
+              if (updated.getOwner().equals(key.getOwner())) {
+                // stay on this screen
+                loadSearches();
+                super.onSaveSuccess(updated, linkName);
+                return;
+              }
+            }
+            History.newItem(Dispatcher.toSearches(updated.getType(),
+              linkName));
+          }
+        };
+      }
+    };
+
+    add(table);
+    add(delete);
+    add(insert);
+  }
+
+  @Override
+  protected void onLoad() {
+    super.onLoad();
+    refresh();
+  }
+
+  public void refresh() {
+    loadSearchList();
+  }
+
+  public void loadSearchList() {
+    Util.SEARCH_LIST_SVC.getSearchList(key.getOwnerId(), cbList);
+  }
+
+  public void loadSearches() {
+    Util.SEARCH_LIST_SVC.getSearches(key.getOwnerId(), cbSearches);
+  }
+
+  private void displaySearches(List<Search> searches) {
+    list.setSearches(searches);
+    table.refresh(key);
+  }
+
+  protected void display(final SearchList list) {
+    this.list = list;
+    OwnerInfo owner = list.getOwnerInfo();
+    key = new Search.Key(owner.getId(), key.get());
+
+    String token = Dispatcher.toSearches(owner.getType(),
+      owner.getOwnerForLink(), key.get());
+    setToken(token);
+
+    displayTitle();
+
+    table.setSearchList(list, key);
+    table.setSavePointerId(token);
+    table.finishDisplay();
+    table.hideEditor();
+
+    insert.setVisible(list.isEditable());
+    delete.setVisible(list.isEditable());
+
+    display();
+  }
+
+  @Override
+  public void registerKeys() {
+    super.registerKeys();
+    table.setRegisterKeys(true);
+  }
+
+  public void onDelete() {
+    delete.setEnabled(false);
+    Util.SEARCH_SVC.deleteSearches(table.getCheckedSearchKeys(), cbDeletes);
+  }
+
+  protected void displayTitle() {
+    String title = getSearchTitle();
+    setPageTitle(title);
+    setWindowTitle(title);
+  }
+
+  protected String getSearchTitle() {
+    String owner = list.getOwnerInfo().getOwnerForDisplay();
+    if ("".equals(owner)) {
+      switch(key.getType()) {
+        case USER:    return Util.C.titleSearchUser();
+        case GROUP:   return Util.C.titleSearchGroup();
+        case PROJECT: return Util.C.titleSearchProject();
+        case SITE:    return Util.C.titleSearchSite();
+      }
+    } else {
+      switch(key.getType()) {
+        case USER:    return Util.M.titleSearchUser(owner);
+        case GROUP:   return Util.M.titleSearchGroup(owner);
+        case PROJECT: return Util.M.titleSearchProject(owner);
+        case SITE:    return Util.C.titleSearchSite();
+      }
+    }
+    return "";
+  }
+}
\ No newline at end of file
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchTable.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchTable.java
new file mode 100644
index 0000000..25b0f5c
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchTable.java
@@ -0,0 +1,290 @@
+// Copyright (C) 2011 The Android Open Source Project
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.client.searches;
+
+import com.google.gerrit.client.Gerrit;
+import com.google.gerrit.client.ui.Hyperlink;
+import com.google.gerrit.client.ui.NavigationTable;
+import com.google.gerrit.common.data.SearchList;
+import com.google.gerrit.reviewdb.Search;
+
+import com.google.gwt.user.client.History;
+import com.google.gwt.user.client.ui.FlexTable.FlexCellFormatter;
+import com.google.gwt.user.client.ui.HTMLTable.Cell;
+import com.google.gwt.user.client.ui.Button;
+import com.google.gwt.user.client.ui.CheckBox;
+import com.google.gwt.user.client.ui.Widget;
+import com.google.gwt.event.dom.client.KeyCodes;
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.event.logical.shared.ValueChangeEvent;
+import com.google.gwt.event.logical.shared.ValueChangeHandler;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class SearchTable extends NavigationTable<Search> {
+  private static int HIDDEN = -1;
+
+  protected SearchEditor editor;
+  private SearchList list;
+  private int editorRow = HIDDEN;
+  private int columns;
+  private boolean inserting = false;
+
+  protected ValueChangeHandler<Boolean> checkVCb =
+      new ValueChangeHandler<Boolean>() {
+        @Override
+        public void onValueChange(ValueChangeEvent<Boolean> event) {
+          if (event.getValue()) {
+            onSearchChecked(true);
+          } else {
+            onSearchChecked(!getCheckedSearchKeys().isEmpty());
+          }
+        }
+      };
+
+  public SearchTable() {
+    keysNavigation.add(new PrevKeyCommand(0, 'k', Util.C.searchListPrev()));
+    keysNavigation.add(new NextKeyCommand(0, 'j', Util.C.searchListNext()));
+    keysNavigation.add(new OpenKeyCommand(0, 'o', Util.C.searchListOpen()));
+    keysNavigation.add(new OpenKeyCommand(0, KeyCodes.KEY_ENTER, Util.C.searchListOpen()));
+
+    table.addClickHandler(new ClickHandler() {
+      @Override
+      public void onClick(ClickEvent event) {
+        final Cell cell = table.getCellForEvent(event);
+        if (cell != null && cell.getCellIndex() != 1
+            && getRowItem(cell.getRowIndex()) != null) {
+          movePointerTo(cell.getRowIndex());
+        }
+      }
+    });
+
+    editor = createSearchEditor();
+  }
+
+  public void setSearchList(SearchList list, Search.Key key) {
+    this.list = list;
+    editor.setSearchList(list);
+    refresh(key);
+  }
+
+  @Override
+  public Search.Key getRowItemKey(Search s) {
+    return s.getKey();
+  }
+
+  @Override
+  public void onOpenRow(int row) {
+    if (row > 0) {
+      movePointerTo(row);
+    }
+  }
+
+  @Override
+  public void movePointerTo(int row) {
+    if (editorRow == (row + 1) && ! inserting) {
+      editor.onCancel();
+    } else {
+      if (editorRow != HIDDEN && editorRow < row) {
+        row--;
+      }
+      hideEditor();
+      super.movePointerTo(row);
+      showEditor(row);
+    }
+  }
+
+  public void movePointerTo(final Search.Key key) {
+    super.movePointerTo(key);
+  }
+
+  protected void showEditor(int row) {
+    hideEditor();
+    if (Gerrit.isSignedIn()) {
+      if (row > 0) {
+        editor.setSearch(getRowItem(row));
+      } else if (row == -1) {
+        row = table.getRowCount() - 1;
+
+        Search.Key key = null;
+        if (list != null && list.getOwnerInfo() != null) {
+          key = new Search.Key(list.getOwnerInfo().getId(), null);
+        }
+        editor.setSearch(new Search(key, null, null));
+
+        inserting = true;
+        clearPointer();
+      }
+      editorRow = row + 1;
+      table.insertRow(editorRow);
+      table.setWidget(editorRow, 0, editor);
+      final FlexCellFormatter fmt = table.getFlexCellFormatter();
+      fmt.setColSpan(editorRow, 0, columns);
+      fmt.addStyleName(editorRow, 0, Gerrit.RESOURCES.css().dataHeader());
+    }
+  }
+
+  protected void hideEditor() {
+    if (editorRow > 0) {
+      table.removeRow(editorRow);
+    }
+    editorRow = HIDDEN;
+    inserting = false;
+  }
+
+  private String link(final Search s) {
+    String type = s.getType().getId();
+    String owner = list.getOwnerInfo().getOwnerForLink();
+    String search = s.getName();
+    if (!"".equals(owner)) {
+      search = owner + "." + search;
+    }
+    return "q," + type + "search:" + search + ",n,z";
+  }
+
+  public void refresh(Search.Key key) {
+    display(list.getSearches());
+    if("" != key.get()) {
+      movePointerTo(key);
+      hideEditor();
+    }
+  }
+
+  protected void display(final List<Search> searches) {
+    hideEditor();
+
+    while (0 < table.getRowCount())
+      table.removeRow(table.getRowCount() - 1);
+
+    populateHeader();
+    if (searches.isEmpty()) {
+      insertNoneRow(1);
+    } else {
+      for (final Search s : searches) {
+        final int row = table.getRowCount();
+        table.insertRow(row);
+        applyDataRowStyle(row);
+        populate(row, s);
+      }
+    }
+  }
+
+  private void insertNoneRow(final int row) {
+    table.insertRow(row);
+    table.setText(row, 0, Util.C.searchTableNone());
+    final FlexCellFormatter fmt = table.getFlexCellFormatter();
+    fmt.setColSpan(row, 0, columns);
+    fmt.setStyleName(row, 0, Gerrit.RESOURCES.css().emptySection());
+  }
+
+  protected void populate(final int row, final Search s) {
+    int col = 1;
+    if (list.isEditable()) {
+      table.setWidget(row, col++, getDeleteBox(row));
+    }
+    table.setWidget(row, col++, new Hyperlink(s.getName(), link(s)));
+    table.setText(row, col++, s.getDescription());
+    table.setText(row, col++, s.getQuery());
+
+    final FlexCellFormatter fmt = table.getFlexCellFormatter();
+    for (col-- ; col >=0 ; col--)
+      fmt.addStyleName(row, col, Gerrit.RESOURCES.css().dataCell());
+
+    col = 1;
+    if (list.isEditable()) {
+      fmt.addStyleName(row, col++, Gerrit.RESOURCES.css().iconCell());
+    }
+    fmt.addStyleName(row, col++, Gerrit.RESOURCES.css().cPROJECT());
+    fmt.addStyleName(row, col++, Gerrit.RESOURCES.css().cSUBJECT());
+    fmt.addStyleName(row, col++, Gerrit.RESOURCES.css().cSUBJECT());
+
+    setRowItem(row, s);
+  }
+
+  public Set<Search.Key> getCheckedSearchKeys() {
+    final Set<Search.Key> keys = new HashSet<Search.Key>();
+    for (int row = 1; row < table.getRowCount(); row++) {
+      Search s = getRowItem(row);
+      if (s != null && table.getWidget(row, 1) instanceof CheckBox
+          && ((CheckBox) table.getWidget(row, 1)).getValue()) {
+        keys.add(s.getKey());
+      }
+    }
+    return keys;
+  }
+
+  public void onSearchChecked(boolean checked) {
+  }
+
+  private Widget getDeleteBox(int row) {
+    CheckBox cb = new CheckBox();
+    cb.addValueChangeHandler(checkVCb);
+    return cb;
+  }
+
+  void populateHeader() {
+    int row = 0;
+    int col = 1;
+    if (list.isEditable()) {
+      table.setText(row, col++, "");
+    }
+    table.setText(row, col++, Util.C.columnSearchName());
+    table.setText(row, col++, Util.C.columnSearchDescription());
+    table.setText(row, col++, Util.C.columnSearchQuery());
+
+    final FlexCellFormatter fmt = table.getFlexCellFormatter();
+    col = 0;
+    fmt.addStyleName(row, col++, Gerrit.RESOURCES.css().iconHeader());
+    if (list.isEditable()) {
+      fmt.addStyleName(row, col++, Gerrit.RESOURCES.css().dataHeader());
+    }
+    fmt.addStyleName(row, col++, Gerrit.RESOURCES.css().dataHeader());
+    fmt.addStyleName(row, col++, Gerrit.RESOURCES.css().dataHeader());
+    fmt.addStyleName(row, col++, Gerrit.RESOURCES.css().dataHeader());
+    columns = col;
+  }
+
+  protected SearchEditor createSearchEditor() {
+    return new Editor(null);
+  }
+
+  protected class Editor extends SearchEditor {
+    public Editor(Search s) {
+      super(s);
+    }
+
+    @Override
+    public void onCancel() {
+      super.onCancel();
+      hideEditor();
+    }
+  }
+}
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/Util.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/Util.java
new file mode 100644
index 0000000..a5f9730
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/Util.java
@@ -0,0 +1,64 @@
+// Copyright (C) 2011 The Android Open Source Project
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.client.searches;
+
+import com.google.gerrit.common.data.SearchListService;
+import com.google.gerrit.common.data.SearchService;
+import com.google.gerrit.reviewdb.Owner;
+import com.google.gwt.core.client.GWT;
+import com.google.gwtjsonrpc.client.JsonUtil;
+
+public class Util {
+  public static final SearchConstants C = GWT.create(SearchConstants.class);
+  public static final SearchMessages M = GWT.create(SearchMessages.class);
+
+  public static final SearchService SEARCH_SVC;
+  public static final SearchListService SEARCH_LIST_SVC;
+
+  static {
+    SEARCH_SVC = GWT.create(SearchService.class);
+    JsonUtil.bind(SEARCH_SVC, "rpc/SearchService");
+
+    SEARCH_LIST_SVC = GWT.create(SearchListService.class);
+    JsonUtil.bind(SEARCH_LIST_SVC, "rpc/SearchListService");
+  }
+
+  public static String getType(Owner.Id id) {
+    return getType(id.getType());
+  }
+
+  public static String getType(Owner.Type type) {
+    switch(type) {
+      case USER:    return C.ownerTypeUser();
+      case GROUP:   return C.ownerTypeGroup();
+      case PROJECT: return C.ownerTypeProject();
+      case SITE:    return C.ownerTypeSite();
+    }
+    return "";
+  }
+}
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/HintTextBox.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/HintTextBox.java
index 6575b61..93b072e 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/HintTextBox.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/HintTextBox.java
@@ -139,7 +139,7 @@ public class HintTextBox extends NpTextBox {
 
   private void onKey(int key) {
     if (key == KeyCodes.KEY_ESCAPE) {
-      setText(prevText);
+      super.setText(prevText);
 
       Widget p = getParent();
       if (p instanceof SuggestBox) {
@@ -208,4 +208,8 @@ public class HintTextBox extends NpTextBox {
 
     isFocused = focus;
   }
+
+  public boolean isHintOn() {
+    return hintOn;
+  }
 }
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/NavigationTable.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/NavigationTable.java
index 0d45529..85ab9ed 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/NavigationTable.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/NavigationTable.java
@@ -96,6 +96,9 @@ public abstract class NavigationTable<RowItem> extends FancyFlexTable<RowItem> {
   }
 
   protected void ensurePointerVisible() {
+    if (getRowItem(currentRow) == null) {
+      return;
+    }
     final int max = table.getRowCount();
     int row = currentRow;
     final int init = row;
@@ -153,6 +156,17 @@ public abstract class NavigationTable<RowItem> extends FancyFlexTable<RowItem> {
     currentRow = newRow;
   }
 
+  protected void clearPointer() {
+    if (currentRow != -1) {
+      final CellFormatter fmt = table.getCellFormatter();
+      final Element tr = DOM.getParent(fmt.getElement(currentRow, C_ARROW));
+      UIObject.setStyleName(tr, Gerrit.RESOURCES.css().activeRow(), false);
+      table.setWidget(currentRow, C_ARROW, null);
+      pointer.removeFromParent();
+    }
+    currentRow = -1;
+  }
+
   protected void scrollIntoView(final Element tr) {
     if (!computedScrollType) {
       parentScrollPanel = null;
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/SwitchingSuggestOracle.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/SwitchingSuggestOracle.java
new file mode 100644
index 0000000..5f03e2a
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/SwitchingSuggestOracle.java
@@ -0,0 +1,76 @@
+// Copyright (C) 2011 The Android Open Source Project
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.client.ui;
+
+import com.google.gwt.user.client.ui.SuggestOracle;
+
+/** This simple SuggestOracle takes an array of SuggestOracles
+ *  and proxies all requests to the currently selected oracle,
+ *  which can be changed with the setIndex() method.
+ */
+public class SwitchingSuggestOracle extends SuggestOracle {
+  private int index = 0;
+  private SuggestOracle[] oracles;
+
+  public SwitchingSuggestOracle(SuggestOracle[] oracles) {
+    setOracles(oracles);
+  }
+
+  public boolean isDisplayStringHTML() {
+    return getOracle().isDisplayStringHTML();
+  }
+
+  public void requestSuggestions(SuggestOracle.Request request,
+    SuggestOracle.Callback callback) {
+    getOracle().requestSuggestions(request, callback);
+  }
+
+  public int getIndex() {
+    return index;
+  }
+
+  public void setIndex(int index) {
+    this.index = index;
+  }
+
+  public SuggestOracle getOracle() {
+    return getOracle(getIndex());
+  }
+
+  public SuggestOracle getOracle(int i) {
+    return getOracles()[i];
+  }
+
+  public SuggestOracle[] getOracles() {
+    return oracles;
+  }
+
+  public void setOracles(SuggestOracle[] oracles) {
+    this.oracles = oracles;
+  }
+}
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/TextBoxBaseListener.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/TextBoxBaseListener.java
new file mode 100644
index 0000000..413ff6f
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/TextBoxBaseListener.java
@@ -0,0 +1,121 @@
+// Copyright (C) 2011 The Android Open Source Project
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.client.ui;
+
+import com.google.gwt.event.dom.client.FocusEvent;
+import com.google.gwt.event.dom.client.FocusHandler;
+import com.google.gwt.event.dom.client.KeyDownEvent;
+import com.google.gwt.event.dom.client.KeyDownHandler;
+import com.google.gwt.event.dom.client.KeyPressEvent;
+import com.google.gwt.event.dom.client.KeyPressHandler;
+import com.google.gwt.event.dom.client.MouseUpEvent;
+import com.google.gwt.event.dom.client.MouseUpHandler;
+
+import com.google.gwt.event.shared.GwtEvent;
+import com.google.gwt.user.client.Command;
+import com.google.gwt.user.client.DeferredCommand;
+import com.google.gwt.user.client.ui.CheckBox;
+import com.google.gwt.user.client.ui.FocusWidget;
+import com.google.gwt.user.client.ui.ListBox;
+import com.google.gwt.user.client.ui.TextBoxBase;
+
+
+/** Listen To all text changes to a TextBoxBase. */
+public class TextBoxBaseListener implements KeyPressHandler, KeyDownHandler,
+   MouseUpHandler {
+
+  private String old = "";
+
+
+  public TextBoxBaseListener(final TextBoxBase tb) {
+    old = tb.getText();
+    tb.addKeyPressHandler(this);
+
+    // Is there another way to capture middle button X11 pastes in browsers
+    // which do not yet support ONPASTE events (Firefox)?
+    tb.addMouseUpHandler(this);
+
+    // Resetting the "original text" on focus ensures that we are
+    // up to date with non-user updates of the text (calls to
+    // setText()...) and also up to date with user changes which
+    // occured after enabling "widget".
+    tb.addFocusHandler(new FocusHandler() {
+        @Override
+        public void onFocus(FocusEvent event) {
+          old = tb.getText();
+        }
+      });
+
+    // CTRL-V Pastes in Chrome seem only detectable via BrowserEvents or
+    // KeyDownEvents, the latter is better.
+    tb.addKeyDownHandler(this);
+  }
+
+  // Handlers
+
+  @Override
+  public void onKeyPress(final KeyPressEvent e) {
+    on(e);
+  }
+
+  @Override
+  public void onKeyDown(final KeyDownEvent e) {
+    on(e);
+  }
+
+  @Override
+  public void onMouseUp(final MouseUpEvent e) {
+    on(e);
+  }
+
+  private void on(final GwtEvent e) {
+    onTextBoxBase((TextBoxBase) e.getSource());
+  }
+
+  private void onTextBoxBase(final TextBoxBase tb) {
+    if (! tb.getText().equals(old)) {
+      onTextChange(old, tb.getText());
+      old = tb.getText();
+    } else {
+
+      // The text appears to not always get updated until the handlers complete.
+      DeferredCommand.add(new Command() {
+        @Override
+        public void execute() {
+          if (! tb.getText().equals(old)) {
+            onTextChange(old, tb.getText());
+            old = tb.getText();
+          }
+        }
+      });
+    }
+  }
+
+  public void onTextChange(String old, String val) {
+  }
+}
-- 
1.7.4.1

